ppx-lib/gen_js_api_ppx.ml

```diff
@@ -1284,8 +1287,43 @@ and gen_funs ~global_attrs p =
     ) p.ptype_params
   in
   let loc = p.ptype_loc in
-  let typvar_used, of_js, to_js =
+  let typvar_used, of_js, to_js, custom_funs =
     match p.ptype_kind with
+    | _ when has_attribute "js.custom" global_attrs ->
+        let result =
+          match get_attribute "js.custom" global_attrs with
+          | Some (k, PStr structure) ->
+              let assert_function label (vbs: value_binding list) =
+                let check vb =
+                  match vb.pvb_pat.ppat_desc with
+                  | Ppat_var nameloc when nameloc.txt = label -> true
+                  | _ -> false
+                in
+                if List.exists check vbs then ()
+                else raise (Error (k.loc, Missing_requried_definitions label))
+              in
+              let rec get_value_bindings (s: structure) =
+                List.concat_map (fun (si: structure_item) ->
+                  match si.pstr_desc with
+                  | Pstr_value (_, vbs) -> vbs
+                  | Pstr_include incdec ->
+                    let result =
+                      match incdec.pincl_mod.pmod_desc with
+                      | Pmod_structure s' -> get_value_bindings s'
+                      | _ -> []
+                    in result
+                  | _ -> []
+                ) s
+              in
+              let vbs = get_value_bindings structure in
+              assert_function (name ^ "_of_js") vbs;
+              assert_function (name ^ "_to_js") vbs;
+              (fun _ -> true),
+              lazy (raise (Error (loc, Union_without_discriminator))),
```

Comment:
I don't have the current logic fresh in my mind, but this looks suspicious : is that lazy ever going to be forced?

Reply:
Yes. It's basically to disable the automatic generation of `_to_js` and `_of_js`. See: https://github.com/LexiFi/gen_js_api/blob/17afcd648b1616badcbaebef561c0a445eaa6641/ppx-lib/gen_js_api_ppx.ml#L1372-L1373

We already have `value_binding`s defined by user, so we don't want to use `push_fun` and just use them as they are.
https://github.com/LexiFi/gen_js_api/blob/17afcd648b1616badcbaebef561c0a445eaa6641/ppx-lib/gen_js_api_ppx.ml#L1406-L1410

But the name of the exception we are using might be confusing. It might be better to add a new one.

Reply:
Ah, yes.  It's better to introduce another exception constructor for that purpose.